### DISCUSSION TITLE
Member 상속 구조 관련 논의 / 회원가입시 닉네임 입력받는 Flow

### DISCUSSION THREAD

**Discussion by her0807** [woowacourse-teams/2022-ternoko#99] [Q&A] [answered]
### 문제점

현재 컬럼 상태는 아래과 같이 필드에 `name 이 없고, nickName 만 있음.`

```java

public class Member {

    @Id
    @GeneratedValue(strategy = IDENTITY)
    private Long id;

    @Column(length = 30)
    private String nickname;

    @Column(nullable = false, unique = true)
    private String email;

    @Column
    private String imageUrl;
}

```

어제 정한 코치, 크루 단건 조회 Response 로  Name 이 들어가야하고, 

```JSON

HTTP/1.1 200 OK
Content-Type: application/json
{
	{
                  "id" : 1,
		"name" : "사현빈",
		"nickname" : "바니",
		"email" : "shb1833@gmail.com",
		"imageUrl" : "imageUrl"
  }
}

```

slack login 시도시 최초 회원가입인지 아닌지 구별하기위한 컬럼 IsSinup 도 없음

```JSON

HTTP/1.1 200 OK
Content-Type: application/json

{
	"isSignup" : true,
	"memberRole" : "COACH",
	"accessToken" : "dsaf.afsdwrq.rewq"
}


```

### 해결방안 

rootEntity 인 Member 에 
- name
- isSignup

두개 추가하면 됨. 

### 생각해볼 내용

회원가입 플로우를 위해서 isSignup 컬럼을 추가하게 되면 최초 회원가입시만 사용하고 버려지는 데이터인데
select * from crew, coach 했을 때마다 데이터를 전체 다 가져오고, 버려지게 된다. 효율적인가 ? 

### 생각해볼 내용에 대한 내 생각.
- 회원 가입로직을 위한 tables 을 빼는게 어떤지 ?
![스크린샷 2022-07-27 오전 9 40 41](https://user-images.githubusercontent.com/26570275/181135920-a95dc039-982c-45a9-afa3-b9cb6d214be5.png)

- 최초의 회원가입할 때 회원가입 시키지말고, userEmail 돌려주고, 닉네임 입력받는 창에서 전체 데이터 받아올 때 save 치기 
아래에 설명 !

###  1️⃣ 최초 로그인 요청시 response 응답값에  email 보냄.

```JSON

HTTP/1.1 200 OK
Content-Type: application/json

{
	"isSignup" : false,
        "email" : temp@gmail.com 
	"memberRole" : "CREW",
	"accessToken" : "dsaf.afsdwrq"
}

```

###  2️⃣  프론트에서 닉네임 입력받는 창 폼에 email 추가하고,nickname 받아서 다시 request 시 save 치기

```JSON
PATCH /api/crews/me  HTTP/1.1
Accept: application/json
Authorization : Bearer ZXCVdfvoasidvhjoqw~~~
{
	{
		"nickname" : "바니",
                "email" : "temp@gmail.com"
		"imageUrl" : "imageUrl"
  }
}

```

**Comment by HyeonbinSa** [comment, marked as answer]
고민 많이 하셨네요.
말씀해주신 내용에 대해 생각해보았는데 'isSignup'을 Entity가 아닌 DTO에만 가지고 있으면 컬럼으로 낭비되지않고 회원가입 테이블을 하나 만들지 않아도 될거같습니다. 
닉네임 기입여부 컬럼이 isSignup이랑 비슷하다고 생각이 들어서 기존 테이블 구조대로 가되 최초 로그인이라면 리스폰스 DTO에 불린 값으로 반환해주면 좋을것같습니다.

**Reply by soominsohn** [reply]
저도 바니 의견에 동의합니다!

**Comment by lokba** [comment]
1️⃣ 최초 로그인 요청시 response 응답값에 email 보냄.

```
HTTP/1.1 200 OK
Content-Type: application/json

{
	"isSignup" : false,
        "email" : temp@gmail.com 
	"memberRole" : "CREW",
	"accessToken" : "dsaf.afsdwrq"
}
```

여기서, imageUrl이 없는데 깜빡하신거죠? 🤔
